internal/plugin/admission/certificaterequest/approval/certificaterequest_approval.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package approval
+
+// CertificateRequestApproval is a plugin that ensures entities that are attempting to
+// modify `status.conditions[type="Approved"]` or `status.conditions[type="Denied"]`
+// have permission to do so (granted via RBAC).
+// Entities will need to be able to `approve` (verb) `signers` (resource type) in
+// `cert-manager.io` (group) with the name `<issuer-type>.<issuer-group>/[<certificaterequest-namespace>.]<issuer-name>`.
+// For example: `issuers.cert-manager.io/my-namespace.my-issuer-name`.
+// A wildcard signerName format is also supported: `issuers.cert-manager.io/*`.
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/jetstack/cert-manager/internal/apis/certmanager"
+	"github.com/jetstack/cert-manager/internal/apis/certmanager/validation/util"
+	"github.com/jetstack/cert-manager/pkg/webhook/admission"
+	"github.com/jetstack/cert-manager/pkg/webhook/admission/initializer"
+)
+
+const PluginName = "CertificateRequestApproval"
+
+type certificateRequestApproval struct {
+	*admission.Handler
+
+	authorizer authorizer.Authorizer
+	discovery  discovery.DiscoveryInterface
+
+	// resourceCache stores the associated APIResource for a given GroupKind
+	// to making multiple queries to the API server for every approval.
+	resourceCache map[schema.GroupKind]metav1.APIResource
```

Comment:
With (2), I always figured it's ultimately a bit of a bug anyway that we can't make an approval decision unless we know if the resource is namespaced. I'd opted to not timeout because any short value is probably not that useful, and any long values is probably also going to be unhelpful if someone _is_ doing (1) (and also makes it unclear to users whether something will or won't succeed, depending on the age of items in this cache).

I'm _inclined_ to actually just opt to cache forever, as it's predictable at the very least. But open to changing it.

> to have odd situations where resource status's reflect impossible states in the current implementation.

It's only approval that really deals with this state, and actually the user is still required to have appropriate RBAC rules configured that *does* grant them access. So it doesn't seem like a totally impossible state... 🤷‍♂️

Reply:
Yeah I guess the impossible state is that an approval condition can be set when there is no signer resource that exists.

> I'm inclined to actually just opt to cache forever, as it's predictable at the very least. But open to changing it.

Indeed, just "feels" wrong. I wonder if there are other examples out there that have tackled similar problems to this.

FWIW, I don't think this is a huge deal, but should at least be documented (outside of code).

Reply:
> Yeah I guess the impossible state is that an approval condition can be set when there is no signer resource that exists.

👍 I don't think we ever really said that this *isn't* possible (i.e. and in the world of Kubernetes CSR objects, the `signerName` is a totally opaque string anyway that doesn't reference any particular 'object')